### PR TITLE
[SPARK-21560][Core] Add hold mode for the LiveListenerBus

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?
1. Add config for hold strategy and the idle capacity.
2. Hold the post method while event queue is full.
3. Notify all holding thread while event queue empty rate lager than the configuration.

## How was this patch tested?

Add a new test in SparkListenerSuite

### CODE REVIEW

**Comment by xuanyuanking** [comment]
ping @gatorsmile @cloud-fan , can you review about this? Thanks :)

**Comment by gatorsmile** [comment]
cc @jiangxb1987

**Comment by xuanyuanking** [comment]
@jiangxb1987 Hi xingbo, can you give me some advise about this?

**Comment by jiangxb1987** [comment]
I’ll review it this week. Thanks!

**Comment by cloud-fan** [comment]
is this still valid? Now `LiveListenerBus` have multiple queues and it's very unlikely the system queues will block.

**Comment by xuanyuanking** [comment]
The hold mode is still valid, I resolved the conflict and add the logic into `AsyncEventQueue`, it can confirm by the test case added in this [patch](https://github.com/apache/spark/pull/18760/files#diff-6ddec7f06d0cf5392943ecdb80fcea24R515).
`Now LiveListenerBus have multiple queue`
Yep, glad to see SPARK-18838 finally merged and it may resolve the event drop problem. But as you say, `it's very unlikely` but may be a hold mode is more safety?

**Comment by cloud-fan** [comment]
cc @vanzin

**Comment by vanzin** [comment]
I'm not super sure this feature will be very useful with queues.

With queues, critical listeners (e.g. dynamic allocation) are very unlikely to miss events. UI listeners are unlikely to miss events. That leaves the event log, which can miss events more often in a large application, but how big of an issue is that?

Also I don't understand why all the complicated code is needed. Why isn't hold mode just calling `queue.put` (blocking) instead of `queue.offer` (non-blocking)?

Long story short, I'd be sort of ok if the code was a lot simpler. But the way it is, it sounds like too much complication for something that I don't really expect to be that big of a problem anymore.

**Comment by xuanyuanking** [comment]
@vanzin Hi Vanzin, thanks a lot for your comments.
Firstly answer your question about `Why isn't hold mode just calling queue.put (blocking) instead of queue.offer (non-blocking)?`
In general scenario of the queue is full, we need a time that hold all events push into the queue, here I use offer to control the `empty rate` in the configuration. If here use `put(blocking)`, this will not relief the queue blocking, and just hanging each post events.

Actually this patch is a internal fix patch for the event dropping problem in Baidu internal env as I described in [jira](https://issues.apache.org/jira/browse/SPARK-21560), glad to see SPARK-18838 has been merged. 

Here I had another thought -- how about I port SPARK-18838 to our product env which solved the event dropping by the current patch, if it works well I just close this. What do you two think about this? @cloud-fan @vanzin :)

**Comment by cloud-fan** [comment]
yea it will be great to see how SPARK-18838 works in real env!

**Comment by vanzin** [comment]
> I use offer(non-blocking) to control the empty rate in the configuration.

Won't that just block all posts until you hit the empty rate? I don't see a lot of difference.

But yeah, it would be good to test how things work with queues before adding yet more features to this part of the code.

**Comment by vanzin** [comment]
So, any verdict on whether this should be closed or not?

**Comment by cloud-fan** [comment]
I think SPARK-18838 should fix the issue, let's close it.

**Comment by xuanyuanking** [comment]
The porting work of SPARK-18838 is in progress based on Spark 2.1 in our product, I'll close this now and open another PR if needed. Thanks @cloud-fan @vanzin
